sample_size/sample_size_calculator/sample_size_calculator.py

```diff
@@ -0,0 +1,105 @@
+from typing import Any
+
+import numpy as np
+from statsmodels.stats.power import NormalIndPower
+from statsmodels.stats.power import TTestIndPower
+
+from sample_size.sample_size_calculator.variables import Boolean
+from sample_size.sample_size_calculator.variables import Numeric
+from sample_size.sample_size_calculator.variables import Ratio
+
+DEFAULT_ALPHA = 0.05
+DEFAULT_POWER = 0.8
+
+
+class PowerAnalysisParameters:
+    """
+    Class to house parameters for power analysis
+    alpha: statistical significance
+    power: statistical power
+    mde: minimum detectable effect
+    """
+
+    mde: float
+    alpha: float = DEFAULT_ALPHA
+    power: float = DEFAULT_POWER
+
+
+class PowerAnalysisType:
+    """
+    Class to house power analysis methods
+    ZTestPower: power analysis method based on Z-test (used by Boolean variable)
+    TTestPower: power analysis method based on T-test (used by Numeric and Ratio variables)
+    """
+
+    ZTestPower = NormalIndPower
+    TTestPower = TTestIndPower
+
+
+class BaseSampleSizeCalculator:
+    def __init__(
+        self, variable_variance: float, variable_analysis_type: Any, power_analysis_parameters: PowerAnalysisParameters
+    ):
+        self.variable_variance = variable_variance
+        self.variable_analysis_type = variable_analysis_type
+        self.power_analysis_parameters = power_analysis_parameters
+        self.std_effect_size = self.get_std_effect_size()
+
+    def get_std_effect_size(self) -> float:
+        return self.power_analysis_parameters.mde / float(np.sqrt(self.variable_variance))
+
+    def get_base_sample_size(self) -> float:
+
+        # perform power analysis to find sample size for given effect
+        power_analysis = self.variable_analysis_type
+        power_analysis_obj = power_analysis()
+
+        sample_size = float(
+            power_analysis_obj.solve_power(
+                effect_size=self.std_effect_size,
+                alpha=self.power_analysis_parameters.alpha,
+                power=self.power_analysis_parameters.power,
+                ratio=1,
+                alternative="two-sided",
+            )
+        )
+
+        return sample_size
+
+
+class SampleSizeCalculators:
```

Comment:
how about making it in a way that doesn't require a lot of refactoring once we support multiple testing (support multiple metrics)? for both the package and the usage. some suggestions below. With that we don't need to refactor much later on.
## Package
```python
class SampleSizeCalculator(object):

    def __init__(self, alpha, power, comparison=2):
        self.alpha = alpha
        self.power = power
        self.comparison = comparison # just for future use, don't need to be included now
        self.boolean_metrics: List[Boolean] = []
        self.numeric_metrics: List[Numeric] = []
        self.ratio_metrics: List[Ratio] = []

    
    def register_bool_metric(self, mde, probability):
        variable = Boolean(mde, probability)
        self.boolean_metrics.append(variable)
    
    def register_ratio_metric(self, mde, numerator_mean, numerator_variance, denominator_mean, denominator_variance, covariance):
        variable = Ratio(mde, numerator_mean, numerator_variance, denominator_mean, denominator_variance, covariance)
        self.ratio_metrics.append(variable)
    
    def register_numeric_metric(self, mde, mean, variance):
        variable = Numeric(mde, mean, variance)
        self.numeric_metrics.append(variable)
    
    def get_overall_sample_size(self):
        # temporarily directly return after calculating the non-null one since it only support single metric now
        if self.boolean_metrics:
            for metric in self.boolean_metrics:
                return self.__get_single_sample_size(metric.variance, metric.mde, NormalIndPower)
        if self.numeric_metrics:
            for metric in self.numeric_metrics:
                return self.__get_single_sample_size(metric.variance, metric.mde, TTestIndPower)
        if self.ratio_metrics:
            for metric in self.ratio_metrics:
                return self.__get_single_sample_size(metric.variance, metric.mde, TTestIndPower)
        
        # In the future, this function will use self.comparison to apply some corrections
        # and eventually return an overall sample size
        # e.g return self.correct_sample_size(sizes, self.comparisoin)

    def __get_single_sample_size(self, mde, variance, power_analysis_type):
        effect_size = mde / float(np.sqrt(variance))
        sample_size = power_analysis_type().solve_power(effect_size, self.alpha, self.power, ratio=1, alternative="two-sided",)
        return sample_size
```

## Usage
```python
calculator = SampleSizeCalculator(alpha, power, comparison)
for metric in metrics:
    if isinstance(metric, Boolean):
        calculator.register_bool_metric(metric.mde, metric.value)
    elif isinstance(metric, Numeric):
        calculator.register_numeric_metric(metric.mde, metric.value, metric.variance)
    elif isinstance(metric, Ratio):
        calculator.register_ratio_metric(...)
sample_size = calculator.get_overall_sample_size()
# Note: the Boolean, Numeric, Ratio here 
# are different than the ones setup in this python package
```